File: packages/sdk-py/dna/kernel/tool_registry.py
"""ToolRegistry — the kernel's @tool definition registry, extracted from the
Kernel god-object (s-kernel-decompose-god-object / kernel-decompose-continue).

Behavior-preserving: the registry dict + the register / lookup / group-filter
logic move verbatim from ``Kernel``; the kernel keeps ``tool`` / ``get_tool`` /
``get_tools`` / ``list_tool_groups`` as thin public delegators (every call site
unchanged), plus a read-only ``_tools`` property for any code that historically
read the dict directly. Tools are global (not tenant-scoped), so one registry is
safely shared across ``with_tenant`` shallow copies.
"""
from __future__ import annotations

from typing import TYPE_CHECKING

if TYPE_CHECKING:  # pragma: no cover
    from dna.kernel.protocols import ToolDefinition


class ToolRegistry:
    """Name → ToolDefinition registry with group-aware filtering."""

    def __init__(self) -> None:
        self._tools: dict[str, "ToolDefinition"] = {}

    def register(self, td: "ToolDefinition") -> None:
        """Register a tool definition. Last-write-wins on same name (idempotent —
        factory-pattern tools re-register on every ``make_*_tools(holder)`` call)."""
        self._tools[td.name] = td

    def get(self, name: str) -> "ToolDefinition | None":
        """Return a tool definition by name, or None if unknown."""
        return self._tools.get(name)

    def get_many(
        self,
        *,
        group: str | None = None,
        groups: "list[str] | set[str] | None" = None,
    ) -> "list[ToolDefinition]":
        """Return registered tool definitions, optionally filtered.

        - ``group="cognitive"`` — exactly that group
        - ``groups=["cognitive", "manifest"]`` — union of groups
        - ``groups=["read"]`` — expands via the 'read' umbrella alias
        Pass nothing to get the full catalog.
        """
        if group is None and not groups:
            return list(self._tools.values())
        from dna.kernel.tools import expand_group_aliases  # noqa: PLC0415
        wanted = {group} if group else set()
        if groups:
            wanted |= set(groups)
        wanted = expand_group_aliases(wanted)
        return [td for td in self._tools.values() if td.group in wanted]

    def groups(self) -> dict[str, list[str]]:
        """Reverse-build {group: [tool_names...]} from the registry."""
        out: dict[str, set[str]] = {}
        for td in self._tools.values():
            if td.group:
                out.setdefault(td.group, set()).add(td.name)
        return {g: sorted(names) for g, names in out.items()}

File: packages/sdk-ts/src/kernel/tool-registry.ts
/**
 * ToolRegistry — the kernel's tool-definition registry. TS twin of the Py
 * `dna.kernel.tool_registry.ToolRegistry`, including READ_UMBRELLA_GROUPS /
 * expand_group_aliases. Landed with s-dna-port-surface-parity (closes the
 * "TS kernel has no tool registry" gap the ExtensionHost contract had to
 * footnote).
 *
 * Both sides are now symmetric: registrants construct `ToolDefinition`s
 * directly and call `kernel.tool(td)` inside `Extension.register()`. (The Py
 * `@dna_tool` langchain decorator that once harvested definitions at
 * function-definition time was removed as dead code — it decorated nothing.)
 *
 * Tools are global (not tenant-scoped), so one registry is safely shared
 * across `withTenant` shallow copies — same posture as the Py twin.
 */

import type { ToolDefinition } from "./protocols.js";

/**
 * 'read' umbrella group — not a real group on any tool; an alias that
 * expands to {code, manifest, docs, eval} at filter time. 1:1 with the Py
 * `READ_UMBRELLA_GROUPS`.
 */
export const READ_UMBRELLA_GROUPS: ReadonlySet<string> = new Set([
  "code", "manifest", "docs", "eval",
]);

/**
 * Expand the 'read' umbrella into its constituent groups. Other group
 * names pass through unchanged. 1:1 with the Py `expand_group_aliases`.
 */
export function expandGroupAliases(
  groups?: Iterable<string> | null,
): Set<string> {
  const out = new Set<string>();
  if (!groups) return out;
  for (const g of groups) {
    if (g === "read") {
      for (const u of READ_UMBRELLA_GROUPS) out.add(u);
    } else {
      out.add(g);
    }
  }
  return out;
}

/** Name → ToolDefinition registry with group-aware filtering. */
export class ToolRegistry {
  private readonly _tools = new Map<string, ToolDefinition>();

  /** Register a tool definition. Last-write-wins on same name (idempotent —
   *  factory-pattern registrants may re-register on every factory call). */
  register(td: ToolDefinition): void {
    this._tools.set(td.name, td);
  }

  /** Return a tool definition by name, or `null` if unknown. */
  get(name: string): ToolDefinition | null {
    return this._tools.get(name) ?? null;
  }

  /**
   * Return registered tool definitions, optionally filtered.
   *
   * - `{ group: "cognitive" }` — exactly that group
   * - `{ groups: ["cognitive", "manifest"] }` — union of groups
   * - `{ groups: ["read"] }` — expands via the 'read' umbrella alias
   *
   * Pass nothing to get the full catalog. 1:1 with the Py `get_many`.
   */
  getMany(opts: {
    group?: string | null;
    groups?: Iterable<string> | null;
  } = {}): ToolDefinition[] {
    const group = opts.group ?? null;
    const extra = opts.groups ? [...opts.groups] : [];
    if (group === null && extra.length === 0) {
      return [...this._tools.values()];
    }
    const wanted = expandGroupAliases([
      ...(group !== null ? [group] : []),
      ...extra,
    ]);
    return [...this._tools.values()].filter(
      (td) => td.group !== null && wanted.has(td.group),
    );
  }

  /** Reverse-build `{group: [toolNames…]}` (names sorted) from the registry. */
  groups(): Record<string, string[]> {
    const out: Record<string, string[]> = {};
    for (const td of this._tools.values()) {
      if (!td.group) continue;
      (out[td.group] ??= []).push(td.name);
    }
    for (const names of Object.values(out)) names.sort();
    return out;
  }
}
